### DISCUSSION TITLE
I don't have access to external dependency in my Extension

### DISCUSSION THREAD

**Discussion by MrlnHi** [Minestom/Minestom#626] [Help] [answered]
I want to use a mysql database, thus I am including an external dependency for that in my extension.json:
```
  "externalDependencies": {
    "repositories": [
      {"name": "Central", "url": "https://repo1.maven.org/maven2/"}
    ],
    "artifacts": [
      "mysql:mysql-connector-java:8.0.26"
    ]
  }
```

however I get a NoClassDefFoundError when trying to run this line:
```
Class<Driver> driverClass = Driver.class;
```

here's the error message:
```
Exception in thread "main" java.lang.NoClassDefFoundError: com/mysql/jdbc/Driver
	at de.mrln.pvpExtension.PvpExtension.initialize(PvpExtension.java:73)
	at java.base/java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:647)
	at net.minestom.server.extensions.ExtensionManager.gotoInit(ExtensionManager.java:130)
	at net.minestom.server.ServerProcessImpl.start(ServerProcessImpl.java:226)
	at net.minestom.server.MinecraftServer.start(MinecraftServer.java:375)
	at net.minestom.server.MinecraftServer.start(MinecraftServer.java:380)
	at TestServer.main(TestServer.java:42)
Caused by: java.lang.ClassNotFoundException: com.mysql.jdbc.Driver
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 7 more
Caused by: java.lang.ClassNotFoundException: com.mysql.jdbc.Driver
```

in my initialize method (I don't want to actually have that line once it works, it was just an easy way for me to reproduce it not working)
Also it works perfectly fine if I use
```
implementation 'mysql:mysql-connector-java:8.0.26'
```
in my build.gradle and remove the externalDependencies part of my extension.json
however it doesn't work if I use
```
compileOnly 'mysql:mysql-connector-java:8.0.26'
```
also I'm using commit `5b699e0`
also also the dependency gets downloaded into `extensions/.libs`

**Comment by jorisguffens** [comment]
I use an ORM in some of my extensions and I need to manually load the driver class before intializing the ORM to make it work. Like this:

```
Class.forName("com.mysql.cj.jdbc.Driver");
```

Also `com.mysql.jdbc.Driver` is deprecated, you should use `com.mysql.cj.jdbc.Driver`

**Reply by MrlnHi** [reply]
Thanks, I'll try that line when I'm home

**Reply by MrlnHi** [reply]
welp, that gives this exception, so no progress
```
java.lang.ClassNotFoundException: com.mysql.cj.jdbc.Driver
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at de.mrln.pvpExtension.PvpExtension.initialize(PvpExtension.java:74)
	at java.base/java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:647)
	at net.minestom.server.extensions.ExtensionManager.gotoInit(ExtensionManager.java:130)
	at net.minestom.server.ServerProcessImpl.start(ServerProcessImpl.java:226)
	at net.minestom.server.MinecraftServer.start(MinecraftServer.java:375)
	at net.minestom.server.MinecraftServer.start(MinecraftServer.java:380)
	at TestServer.main(TestServer.java:42)
Exception in thread "main" java.lang.NoClassDefFoundError: com/mysql/cj/jdbc/Driver
	at de.mrln.pvpExtension.PvpExtension.initialize(PvpExtension.java:78)
	at java.base/java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:647)
	at net.minestom.server.extensions.ExtensionManager.gotoInit(ExtensionManager.java:130)
	at net.minestom.server.ServerProcessImpl.start(ServerProcessImpl.java:226)
	at net.minestom.server.MinecraftServer.start(MinecraftServer.java:375)
	at net.minestom.server.MinecraftServer.start(MinecraftServer.java:380)
	at TestServer.main(TestServer.java:42)
Caused by: java.lang.ClassNotFoundException: com.mysql.cj.jdbc.Driver
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 7 more
Caused by: java.lang.ClassNotFoundException: com.mysql.cj.jdbc.Driver
```

**Reply by MrlnHi** [reply]
As I said, when using implementation on the dependency (which I don't want to because 1st i don't wanna shade it into my extension jar and 2nd there's literally a feature in minestom that's supposed to help) it works.

**Reply by tristan-f-r** [reply]
I can confirm I'm having the same issue in the extension improvement branch with luckperms: https://github.com/Project-Cepi/LuckPerms. @mworzala may be able to help with this.

**Comment by mworzala** [comment, marked as answer]
There are two things in play here I believe, both related to how classes are loaded. I am running under the assumption that you are using the indev classes mechanism here, if not then I cannot replicate the issue.
1. When you call `Class.forName(String)`, the classloader of the current class is used. This is OK in most cases, because typically the extension classloader _is_ the classloader for your extension. indev classes is an exception to this rule (more below). A fix in this case is to do `Class.forName("com.mysql.cj.jdbc.Driver", true, getOrigin().getClassLoader())`.
2. Some background on java classloaders is relevant here. When loading a class, the classloader will first search its parent and then itself. This means that if the class is on the class path of the parent it will be used. In Minestom, the extension classloader has a parent of the root classloader (which contains all of the minestom classes). When using indev classes, your classes exist on the class path generated by gradle (eg your src directory) meaning the extension main is loaded in the app classloader (same as Minestom). Your external dependencies however are being loaded in the extension classloader, which the app classloader cannot access without specifying it (eg 1.).

TL:DR `Class.forName("com.mysql.cj.jdbc.Driver", true, getOrigin().getClassLoader())` is a solution for this specific case, but the greater problem here will be solved in #529.

@LeoDog896 your issue is unrelated. #529 does not have the extension dependency mechanism implemented yet. It is one of the last remaining things. My hope is to finish up that branch this weekend, but I have been crazy busy so we'll see.
